File: packages/grid-core/src/model/GridConfig.ts
export type SelectionMode = 'none' | 'row' | 'multi-row' | 'cell' | 'multi-cell'

export type GridPaginationConfig = {
  enabled?: boolean
  pageSize?: number
  pageIndex?: number
}

export type GridConfig = {
  selection?: {
    mode?: SelectionMode
    checkboxes?: boolean
  }
  rowHeader?: {
    enabled?: boolean
    width?: number
  }
  pagination?: GridPaginationConfig
}

export type ResolvedGridConfig = {
  selection: {
    mode: SelectionMode
    checkboxes: boolean
  }
  rowHeader: {
    enabled: boolean
    width: number
  }
  pagination: {
    enabled: boolean
    pageSize: number
    pageIndex: number
  }
}

function normalizePageSize(value: number | undefined) {
  return Number.isFinite(value) && value! > 0 ? Math.max(1, Math.floor(value!)) : 50
}

function normalizePageIndex(value: number | undefined) {
  return Number.isFinite(value) && value! >= 0 ? Math.floor(value!) : 0
}

export function normalizeConfig(config: GridConfig = {}): ResolvedGridConfig {
  return {
    selection: {
      mode: config.selection?.mode ?? 'none',
      checkboxes: config.selection?.checkboxes ?? false
    },
    rowHeader: {
      enabled: config.rowHeader?.enabled ?? false,
      width: config.rowHeader?.width ?? 56
    },
    pagination: {
      enabled: config.pagination?.enabled ?? false,
      pageSize: normalizePageSize(config.pagination?.pageSize),
      pageIndex: normalizePageIndex(config.pagination?.pageIndex)
    }
  }
}

File: packages/grid-blazor/src/Models/GridConfig.cs
using System.Text.Json.Serialization;

namespace Tipolox.LitGrid.Blazor;

public class GridConfig
{
    [JsonPropertyName("selection")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public GridSelectionConfig? Selection { get; set; }

    [JsonPropertyName("rowHeader")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public GridRowHeaderConfig? RowHeader { get; set; }

    [JsonPropertyName("pagination")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public GridPaginationConfig? Pagination { get; set; }
}

public class GridSelectionConfig
{
    [JsonPropertyName("mode")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public SelectionMode? Mode { get; set; }

    [JsonPropertyName("checkboxes")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public bool? Checkboxes { get; set; }
}

public class GridRowHeaderConfig
{
    [JsonPropertyName("enabled")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public bool? Enabled { get; set; }

    [JsonPropertyName("width")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public double? Width { get; set; }
}

public class GridPaginationConfig
{
    [JsonPropertyName("enabled")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public bool? Enabled { get; set; }

    [JsonPropertyName("pageSize")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public int? PageSize { get; set; }

    [JsonPropertyName("pageIndex")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public int? PageIndex { get; set; }
}
